pfio/MultiGroupServer.F90

```diff
@@ -419,13 +419,15 @@ subroutine receive_output_data(this, rc)
 
         call Mpi_Bcast( back_local_rank, 1, MPI_INTEGER, 0, this%front_comm, ierror)
 
-        if (allocated(this%buffers(back_local_rank+1)%buffer)) call MPI_Wait(this%buffers(back_local_rank+1)%request, MPI_STAT, ierror)
         call f_d_ms(collection_counter)%serialize(this%buffers(back_local_rank+1)%buffer)
+        call f_d_ms(collection_counter)%destroy(_RC)
         msg_size= size(this%buffers(back_local_rank+1)%buffer)
         call Mpi_send(msg_size,1, MPI_INTEGER, this%back_ranks(back_local_rank+1), &
              this%back_ranks(back_local_rank+1), this%server_comm, ierror)
         call Mpi_Isend(this%buffers(back_local_rank+1)%buffer, msg_size, MPI_INTEGER, this%back_ranks(back_local_rank+1), &
             this%back_ranks(back_local_rank+1), this%server_comm, this%buffers(back_local_rank+1)%request,ierror)
+        call MPI_Wait(this%buffers(back_local_rank+1)%request, MPI_STAT, ierror)
```

Comment:
Doesn't placing the WAIT here defeat the asynchronous aspects of the communication pattern?

Reply:
Yes, I believe it slows down the server. That is a tradeoff.  Maybe we should make it smarter. If the memory is not an issue, then we don't make changes here

Reply:
Unless you can measure it for some realistic use cases, let's not slow the code down.   Asynchronicity is more important than memory footprint.